### PR TITLE
[JBWS-4179] WS integration with WildFly Elytron - AuthenticationClien…

### PR DESCRIPTION
**Pull request by Skyllarr**
…t for Authentication / SSL

[EAP7-1097](https://issues.jboss.org/browse/EAP7-1097)

### CODE REVIEW

**Comment by Skyllarr** [comment]
retest this please

**Comment by Skyllarr** [comment]
@jimma it is possible that the CI needs an update. I think this PR should pass the checks

**Comment by jimma** [comment]
Look the travis is problematic to build with oraclejdk8. I changed travis build with openjdk8, please rebase .
Can you please create a jbossws jira instead of use a EAP jira for this new feature ? We only use EAP jira number in production branch. Thanks.

File: src/main/java/org/jboss/wsf/spi/security/ClientConfigProvider.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+public interface ClientConfigProvider {
+
+    String CLIENT_PROVIDER_CONFIGURED = "clientProviderConfigured";
+    String CLIENT_USERNAME = "clientUsername";
+    String CLIENT_PASSWORD = "clientPassword";
+    String CLIENT_HTTP_MECHANISM = "clientHttpMechanism";
+    String CLIENT_WS_SECURITY_TYPE = "clientWSSecurityType";
+    String CLIENT_SSL_CONTEXT = "clientSSLContext";
+
+    ClientConfig configure(ClientConfig config, URI uri) throws ClientConfigException;
```
[the remark this answers is not in the document]
Comment (Skyllarr) [reply src/main/java/org/jboss/wsf/spi/security/ClientConfigProvider.java]: Elytron config is configured by URI and returns name/password/sslcontext accordingly. This way each URI can have different configuration.

File: src/main/java/org/jboss/wsf/spi/metadata/config/ClientConfig.java [hunk 2]
```diff
@@ -36,12 +38,20 @@
...
          List<UnifiedHandlerChainMetaData> postHandlerChains, Map<String, String> properties,
          Map<String, Feature> features)
    {
-      super(configName, preHandlerChains, postHandlerChains, properties, features);
+      this(configName, preHandlerChains, postHandlerChains, properties, features, null);
+   }
+
+   public ClientConfig(String configName, List<UnifiedHandlerChainMetaData> preHandlerChains,
+                       List<UnifiedHandlerChainMetaData> postHandlerChains, Map<String, String> properties,
+                       Map<String, Feature> features, Map<String, Object> additionalProperties)
```
[the remark this answers is not in the document]
Comment (Skyllarr) [reply src/main/java/org/jboss/wsf/spi/metadata/config/ClientConfig.java]: @jimma I need to pass SSLContext via properties, so I need an Object property. Hence Map<String, Object>.

[the remark this answers is not in the document]
Comment (jimma) [reply src/main/java/org/jboss/wsf/spi/metadata/config/ClientConfig.java]: You can change properties to Map<String, Object> , then you don't need another properties.

File: src/main/java/org/jboss/wsf/spi/security/ClientConfigProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jimma) [reply src/main/java/org/jboss/wsf/spi/security/ClientConfigProvider.java]: Can you point me where this uri will be used (in jbossws-cxf) later?

File: src/main/java/org/jboss/wsf/spi/metadata/config/ClientConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Skyllarr) [reply src/main/java/org/jboss/wsf/spi/metadata/config/ClientConfig.java]: interface `CommonConfig` has the following methods that are being used on many places also on server side. 
```java
public String getProperty(String name);
public Map<String, String> getProperties();
```
Should I change that interface everywhere it is being used?

File: src/main/java/org/jboss/wsf/spi/security/ClientConfigProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Skyllarr) [reply src/main/java/org/jboss/wsf/spi/security/ClientConfigProvider.java]: It is not used later it is used to load the config here: 
https://github.com/jbossws/jbossws-cxf/pull/108/commits/40ec1435e982566b95942cbe4d5fde06a17908f4#diff-6f8b4750ba45903674072c139b1f566aR75
and here
https://github.com/jbossws/jbossws-common/pull/18/commits/5679decbd38020edf79bd9739f3b1dbc6aa9ab9a#diff-7bb69e45f9d40fb90f29eb576f680e2fR174

File: src/main/java/org/jboss/wsf/spi/metadata/config/ClientConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jimma) [reply src/main/java/org/jboss/wsf/spi/metadata/config/ClientConfig.java]: Is it doable  that we created another properties with type Map<String,Object> that we passed all info with this properties ?

[the remark this answers is not in the document]
Comment (Skyllarr) [reply src/main/java/org/jboss/wsf/spi/metadata/config/ClientConfig.java]: @jimma As we discussed in chat, I updated PR where I added `Map<String, Object> attachments` where all info from ClientConfigProvider is stored. Other properties stay in `Map<String, String> properties`.

File: src/main/java/org/jboss/wsf/spi/security/ClientConfigProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jimma) [reply src/main/java/org/jboss/wsf/spi/security/ClientConfigProvider.java]: @Skyllarr  Why do we have to add the attachment parameter in the constructor ?

[the remark this answers is not in the document]
Comment (Skyllarr) [reply src/main/java/org/jboss/wsf/spi/security/ClientConfigProvider.java]: @jimma I updated the PRs where I removed attachment from constructor

**Pull request merged**